apps/s_client.c

```diff
@@ -3553,4 +3556,69 @@ static char *base64encode (const void *buf, size_t len)
     return out;
 }
 
+/*
+ * Host dNS Name verifier: used for checking that the hostname is in dNS format 
+ * before setting it as SNI
+ */
+static int is_dNS_name(const char *host)
+{
+    const size_t MAX_LABEL_LENGTH = 63;
+    size_t i;
+    int isdnsname = 0;
+    size_t length = strlen(host);
+    size_t label_length = 0;
+    int all_numeric = 1;
+
+    /*
+     * Deviation from strict DNS name syntax, also check names with '_'
+     * Check DNS name syntax, any '-' or '.' must be internal,
+     * and on either side of each '.' we can't have a '-' or '.'.
+     *
+     * If the name has just one label, we don't consider it a DNS name.
+     */
+    for (i = 0; i < length && label_length < MAX_LABEL_LENGTH; ++i) {
+        char c = host[i];
+
+        if ((c >= 'a' && c <= 'z')
+            || (c >= 'A' && c <= 'Z')
+            || c == '_') {
+            label_length += 1;
+            all_numeric = 0;
+            continue;
+        }
+
+        if (c >= '0' && c <= '9') {
+            label_length += 1;
+            continue;
+        }
+
+        /* Dot and hyphen cannot be first or last. */
+        if (i > 0 && i < length - 1) {
+            if (c == '-') {
+                label_length += 1;
+                continue;
+            }
+            /*
+             * Next to a dot the preceding and following characters must not be
+             * another dot or a hyphen.  Otherwise, record that the name is
+             * plausible, since it has two or more labels.
+             */
+            if (c == '.'
+                && host[i + 1] != '.'
+                && host[i - 1] != '-'
+                && host[i + 1] != '-') {
+                label_length = 0;
```

Comment:
Don't we need to check ```label_length``` against ```MAX_LABEL_LENGTH``` here?  At present I can only see us making that check for the final label (which is probably a TLD anyway).

Reply:
The check is done for all labels - see end condition on the for.

Reply:
So it is, my apologies.  (It seems that I got interrupted mid-review and lost some state.)
That said, the RFC allows a label of length exactly 63, but my first read of the for loop (and this check) would be that it would reject such a label.

Reply:
This does, though, suggest that we may want some test cases with 63- and 64-byte labels.

Reply:
I think attempting to test this would need the test itself to make calls across the network so I'll avoid doing that, but if there's a way to mock said calls I could give it a go - @mattcaswell ?

Reply:
We don't really have a good way to do that at the moment.

Reply:
Looking at the comment by @kaduk I also agree that the code seems to reject 63-octet labels, which are rare, but legal.